test Installations

annacrombie opened the pull request.

kzc commented:
@annacrombie A significant bootstrap/build/test performance regression was introduced in 1ad5ffac7a18cef9fb0d41d46c5f1d190b1b4477:

65.6s commit 40e92d0ea9976e2606b80e36d7a353e4f94bf184 update meson tests json
65.0s commit 1ad5ffac7a18cef9fb0d41d46c5f1d190b1b4477 project tests: support test.json
32.5s commit 5ecf48c4672a3de1dfaa707f8e9b283b5b7871d4 add repr function
32.7s commit 5ecf48c4672a3de1dfaa707f8e9b283b5b7871d4 add repr function

Observed on windows msys2/mingw64.

Can this test.json support be an explicit opt-in?  Or at least provide a way to opt out of it?

annacrombie commented:
@kzc that's really strange.  Rather than make test.json optional, I'd like to know why tests take 2x as long?

annacrombie commented:
@kzc could it be this?
```
+    '-Denv.NINJA=' + repr(ninja),
```

I noticed during testing that the ninja passed to the test runner wasn't always respected, so this should force that ninja to be used in all cases.  Before this change there would be conflicts where muon was built with samu but then muon test invoked ninja which rebuilt everything/

kzc commented:
It's not ninja related. I build and test with `NINJA=samu`. Just to be sure:
```
$ mv /usr/bin/ninja.exe /usr/bin/ninja.exe-
$ ninja --version
-bash: ninja: command not found
```
Same 68s timing.

The odd thing is that there is not a copious amount of json produced compared to the rest of the files:
```
# all build files
$ find build -type f | wc -l
11211

# json build files
$ find build -name '*.json' | wc -l
4302

# top 10 json file sizes in K
$ find build -name '*.json' | xargs -d '\n' du -k | sort -n | tail -10
16	build/tests/project/python/5 modules kwarg/meson-info/intro-buildoptions.json
16	build/tests/project/python/7 install path/meson-info/intro-buildoptions.json
16	build/tests/project/unit/65 alias target/meson-info/intro-buildoptions.json
20	build/tests/project/common/141 special characters/meson-info/intro-targets.json
20	build/tests/project/common/150 reserved targets/meson-info/intro-targets.json
24	build/tests/project/common/145 recursive linking/compile_commands.json
32	build/tests/project/common/145 recursive linking/meson-info/intro-targets.json
136	build/compile_commands.json
612	build/.muon/tests.json
12068	build/tests/project/common/254 long output/.muon/tests.json

# json build files in K
$ find build -name '*.json' | xargs -d '\n' du -k | awk '{n+=$1}END{print n}'
23507

# all build files in K
$ find build -type f | xargs -d '\n' du -k | awk '{n+=$1}END{print n}'
148027
```
4302 json files in `build` totaling 23.5M.
This compares to 148M for all files in `build`.

`rm -rf build` is run before each timing.  Does muon cache any file(s) outside of the build directory in a `$HOME` dot directory or something?

annacrombie commented:
@kzc muon doesn't cache anything outside the build directory.  Also, each project test should rm -rf its output directory so rerunning the project tests even without deleting the whole build folder should always result in similar timings.

I added some basic timing collection data to the tests and added a printed report to the msys2 build.  Here is a result from clang64:
```
eval
  0.0% of total time
  longest: tests/project/common/107 spaces backslash/test_report.dat, 10ms
prerun
  0.0% of total time
  longest: tests/project/common/153 wrap file should not failed/test_report.dat, 272ms
analyze
  3.2% of total time
  longest: tests/project/common/129 build by default/test_report.dat, 649ms
setup
  23.5% of total time
  longest: tests/project/common/103 has header symbol/test_report.dat, 12147ms
build
  32.1% of total time
  longest: tests/project/common/145 recursive linking/test_report.dat, 11730ms
test
  26.4% of total time
  longest: tests/project/common/51 run target/test_report.dat, 8541ms
install
  6.6% of total time
  longest: tests/project/common/53 install script/test_report.dat, 4002ms
check_install
  0.0% of total time
  longest: tests/project/common/12 data/test_report.dat, 102ms
uninstall
  7.8% of total time
  longest: tests/project/common/45 custom install dirs/test_report.dat, 1191ms
```

The install step was already being run, but might be doing more work now that it actually, well, works :).  The uninstall step is new, and it does take an appreciable % of the total time.  I don't think it would be enough for a 2x increase, but you could try commenting out the `step.uninstall` in the project tests runner and see if that fixes it.  It is pretty surprising to me that uninstall takes _that_ long, I'll take another look at it once I have a minute to work on my windows machine.  check_install is the snippet that actually deals with the test.json and it doesn't seem to be an issue.

For comparison, the timings on my mac m2 are around 2% of time for install and 1.5% of time for uninstall, with the vast majority in setup/build/test.

kzc commented:
@annacrombie This patch on latest master 7fd1ba9239251a6fd5c89518a6496afff8d65f1b fixes the test speed regression on msys2. I don't pretend to understand it - I was just studying the regression commit 1ad5ffac7a18cef9fb0d41d46c5f1d190b1b4477 and doing random monkey reversion:

```patch
--- a/tests/project/runner.meson
+++ b/tests/project/runner.meson
@@ -91,9 +91,9 @@ endfunc

 steps.build = func()
-    run_command(ninja, '-C', build, check: true, capture: false)
+    run_command(ninja, '-C', build, check: true)
 endfunc

 steps.test = func()
-    run_command(muon, '-C', build, 'test', '-a', check: true, capture: false)
+    run_command(muon, '-C', build, 'test', '-a', check: true)
 endfunc

@@ -106,5 +106,4 @@ steps.install = func()
         env: {'DESTDIR': destdir},
         check: true,
-        capture: false,
     )
 endfunc
@@ -216,5 +215,4 @@ steps.uninstall = func()
         '-Ud', destdir,
         check: true,
-        capture: false,
     )

```

With AV disabled it takes 38s to bootstrap/build/test with clang or gcc. Not quite the 33s before the regression, but certainly better than 66s. Interestingly with this patch CPU utilization across all 16 virtual cores is nearly 100%. Prior to the patch per core CPU utilization was around 50%.

So what does the presence of this `run_command`  `capture: false` parameter do or not do?

annacrombie commented:
Interesting.  capture: false is supposed to prevent run_command from redirecting the subprocess's stdout and stderr to a pipe.  It is similar to the run_cmd flag `run_cmd_ctx_flag_dont_capture`.  I set capture: false in the tests so that all of the steps (setup, build, test, etc.) will have their output present in the test log.  This can be useful in the case of a failure: imagine a build step fails, if the output from setup was captured (and not subsequently printed) then you won't have the setup output in the test failure output, only the output from the step that failed.

I guess this is slower because if each invocation of `muon internal eval runner.meson` slurps up all its subprocess output and throws it away in the common non-failing case, then the parent `muon test` process has less output to gather from its children.  Since the impact is so severe, it makes sense to revert the capture: false changes.
